### PR TITLE
feat: add hardfork storage to metadata contract

### PR DESCRIPTION
**Pull request by driftluo**
## What this PR does / why we need it?

This PR implemented storage-related functionality in hardfork：

- HARDFORK_KEY to stores all planned or achieved hardfork times
- HARDFORK_INFO caches current hardfork flags
- flags will be H256, Any hardfork will be represented by an independent and non-repeating bit

All hardfork-related functionality must be updated at every block generation. 

Restart, activation, and startup status settings will be improved in subsequent PRs

### What is the impact of this PR?

No Breaking Change

**PR relation**:
- Ref #

<details><summary>CI Settings</summary><br/>

### **CI Usage**

**Tip**: Check the CI you want to run below, and then comment `/run-ci`.

**CI Switch**

- [x] E2E Tests
- [ ] Code Format
- [x] Web3 Compatible Tests
- [ ] OCT 1-5 And 12-15
- [ ] OCT 6-10
- [ ] OCT 11
- [ ] OCT 16-19
- [x] v3 Core Tests

### **CI Description**

| CI Name                                   | Description                                                               |
| ----------------------------------------- | ------------------------------------------------------------------------- |
| *Chaos CI*                                | Test the liveness and robustness of Axon under terrible network condition |
| *Cargo Clippy*                            | Run `cargo clippy --all --all-targets --all-features`                     |
| *Coverage Test*                           | Get the unit test coverage report                                         |
| *E2E Test*                                | Run end-to-end test to check interfaces                                   |
| *Code Format*                             | Run `cargo +nightly fmt --all -- --check` and `cargo sort -gwc`           |
| *Web3 Compatible Test*                    | Test the Web3 compatibility of Axon                                       |
| *v3 Core Test*                            | Run the compatibility tests provided by Uniswap V3                        |
| *OCT 1-5 \| 6-10 \| 11 \| 12-15 \| 16-19* | Run the compatibility tests provided by OpenZeppelin                      |

</details>

### CODE REVIEW

File: protocol/src/types/executor.rs [hunk 1]
```diff
@@ -52,6 +52,7 @@ pub struct ExecutorContext {
     pub gas_price:              U256,
     pub block_gas_limit:        U256,
     pub block_base_fee_per_gas: U256,
+    pub extra_data:             bytes::Bytes,
```
Comment (KaoImin) [inline protocol/src/types/executor.rs]: `use crate::types::Bytes`
```suggestion
    pub extra_data:             Bytes,
```

File: core/executor/src/system_contract/metadata/mod.rs [hunk 2]
```diff
@@ -98,6 +103,23 @@ impl<Adapter: ExecutorAdapter + ApplyBackend> SystemContract<Adapter>
...
+                update_states(adapter, sender, Self::ADDRESS);
+
+                return {
+                    let mut resp = succeed_resp(gas_limit);
+                    resp.ret = info
+                        .unwrap_or(HardforkInfo::default().encode().unwrap())
+                        .to_vec();
+                    resp
+                };
+            }
```
Comment (KaoImin) [inline core/executor/src/system_contract/metadata/mod.rs]: Query the hard fork info by transaction is unnecessary, use jsonrpc instead.

File: core/executor/src/system_contract/metadata/mod.rs [hunk 3]
```diff
@@ -30,6 +33,8 @@ const METADATA_CACHE_SIZE: NonZeroUsize = unsafe { NonZeroUsize::new_unchecked(1
 lazy_static::lazy_static! {
     static ref EPOCH_SEGMENT_KEY: H256 = Hasher::digest("epoch_segment");
     static ref CKB_RELATED_INFO_KEY: H256 = Hasher::digest("ckb_related_info");
+    static ref HARDFORK_KEY: H256 = Hasher::digest("hardfork");
+    static ref HARDFORK_INFO: RwLock<U256> = RwLock::new(U256::zero());
```
Comment (KaoImin) [inline core/executor/src/system_contract/metadata/mod.rs]: `ArcSwap` has a better efficiency than `RwLock` in this situation.

File: core/executor/src/system_contract/metadata/store.rs [hunk 4]
```diff
@@ -155,4 +157,63 @@ impl MetadataStore {
...
+                    if let Some(i) = hardfork_info.inner.last_mut() {
+                        let info = i.flags & info;
+                        if i.block_number == block_number {
+                            i.flags = info;
+                        } else {
+                            hardfork_info.inner.push(HardforkInfoInner {
+                                block_number,
+                                flags: info,
+                            });
+                        }
```
Comment (KaoImin) [inline core/executor/src/system_contract/metadata/store.rs]: Abstract the code as a `fn add(&self, other: HardforkInfoInner)` method of `HardforkInfo`.

File: builtin-contract/metadata/contracts/metadata.sol [hunk 5]
```diff
@@ -148,4 +157,9 @@ contract MetadataManager {
     function setCkbRelatedInfo(
         CkbRelatedInfo memory ckbRelatedInfo
     ) public view {}
+
+    function hardforkInfo() public view returns (HardforkInfo memory) {
+        HardforkInfoInner[] memory inner = new HardforkInfoInner[](0);
+        return HardforkInfo(inner);
+    }
```
Comment (KaoImin) [inline builtin-contract/metadata/contracts/metadata.sol]: Unnecessary code

**Comment by KaoImin** [comment]
The `ABI` of metadata contract does not need to be changed.

File: protocol/src/types/primitive.rs [hunk 6]
```diff
@@ -429,6 +429,37 @@ pub struct CkbRelatedInfo {
...
+
+        self.inner
+            .sort_by(|a, b| a.block_number.cmp(&b.block_number));
+    }
+}
+
+#[derive(RlpEncodable, RlpDecodable, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
+pub struct HardforkInfoInner {
+    pub block_number: BlockNumber,
+    pub flags:        U256,
```
Comment (KaoImin) [inline protocol/src/types/primitive.rs]: The `derive(Serialize)` will make `U256` lose the leading zero. Perhaps `H256` is more applicable.

Comment (driftluo) [reply protocol/src/types/primitive.rs]: I think `#[serde(serialize_with = "serialize_uint"))]` is enough

File: protocol/src/types/primitive.rs [hunk 7]
```diff
@@ -429,6 +431,38 @@ pub struct CkbRelatedInfo {
...
 
+#[derive(
+    RlpEncodable, RlpDecodable, Serialize, Deserialize, Clone, Debug, PartialEq, Eq, Default,
+)]
+pub struct HardforkInfo {
+    pub inner: Vec<HardforkInfoInner>,
+}
+
+impl HardforkInfo {
+    pub fn push(&mut self, mut other: HardforkInfoInner) {
```
Comment (KaoImin) [inline protocol/src/types/primitive.rs:440]: Add unit test for this method.

**Review by KaoImin — DISMISSED** [review]

**Review by KaoImin — DISMISSED** [review]

File: protocol/src/types/primitive.rs (hunk 6 above)
Comment (chaoticlonghair) [reply protocol/src/types/primitive.rs]: I don't understand why the type of this field `flags` is `U256`, not `H256`.

If I'm correct, there was no arithmetic for this field.

Comment (driftluo) [reply protocol/src/types/primitive.rs]: here we need to use “&” operations to ensure that the next hardfork brings the information of all previous hardforks.

https://github.com/axonweb3/axon/pull/1380/files#diff-9e85d32a5e7a6dc441136d66a31b595ef11780890af67a35b58fd72d69a0d950R442

Comment (chaoticlonghair) [reply protocol/src/types/primitive.rs]: > here we need to use “&” operations ...

Ref: [`impl BitAnd<H256> for H256` in `ethereum-types`](
https://docs.rs/ethereum-types/0.14.1/ethereum_types/struct.H256.html#impl-BitAnd%3CH256%3E-for-H256)

p.s. The naming of `serialize_to_h256` is terrible.

p.s.s. Bit-and is not arithmetic.

**Review by chaoticlonghair — APPROVED** [review]

**Review by KaoImin — APPROVED** [review]

**Pull request merged**